src/pint/models/timing_model.py

```diff
@@ -1987,17 +1991,29 @@ def validate_toas(self, toas):
         lists some (at least one) of the problem parameters.
         """
         bad_parameters = []
+        warning_parameters = []
         for maskpar in self.get_params_of_type_top("maskParameter"):
             par = getattr(self, maskpar)
-            if "TNEQ" in str(par.name) or par.frozen:
+            if "TNEQ" in str(par.name):
                 continue
             if len(par.select_toa_mask(toas)) == 0:
-                bad_parameters.append(f"'{maskpar}, {par.key}, {par.key_value}'")
+                msg = f"'{maskpar}, {par.key}, {par.key_value}'"
+                if not par.frozen:
+                    bad_parameters.append(msg)
+                else:
+                    warning_parameters.append(msg)
+
+        # The component level of validate_toas only accepts the bad_parameters,
+        # not warning parameters.
         for c in self.components.values():
             try:
                 c.validate_toas(toas)
             except MissingTOAs as e:
                 bad_parameters += e.parameter_names
+
+        # first print out the warnings
+        if warning_parameters:
+            log.warning(missingTOAs_msg(warning_parameters))
```

Comment:
If you use `warnings.warn` here you can use `pytest.warns` below to verify that warnings are emitted.

Reply:
I was thinking about it, but in the end, I followed the tradition. Just want to know, why does astropy implement its own warning? Is there any reason we should stick with astropy warning?

Reply:
We've had this discussion before, I think.  I think it is just that the astropy logging is much more than warnings.  It has multiple levels (including INFO and DEBUG -- which we should use more often) and has colors and other nice capabilities.  It isn't like it is a new dependency or anything, and we use it all over in PINT.  So I don't think it is worth worrying about.  I'm not sure that the `pytest.warns` is a big enough improvement to cause us to switch.  But I'm happy to hear other arguments.

Reply:
Astropy itself has largely made the switch over to using `warnings`; the `warnings` mechanism can easily be logged using the Astropy logging mechanism. If we use `warnings` we can write tests that verify that we emit the important warnings, and our users can choose what to do with them - for example, use `-Werror` to promote them into exceptions if they want to write code that doesn't emit any. There is discussion of this in #797 ; in one comment there I added:

From https://docs.python.org/3/howto/logging.html :

>    warnings.warn() in library code if the issue is avoidable and the client application should be modified to eliminate the warning
>
>    logging.warning() if there is nothing the client application can do about the situation, but the event should still be noted